File: Real-Time-Query-Workshop/java/code/src/main/java/com/redis/queryworkshop/Lab3.java
/**
 * @fileoverview Basic Search operations with the redis client
 * @maker Joey Whelan
 */
package com.redis.queryworkshop;
import java.util.List;
import com.google.gson.Gson;
import redis.clients.jedis.JedisPooled;
import redis.clients.jedis.search.IndexDefinition;
import redis.clients.jedis.search.IndexOptions;
import redis.clients.jedis.search.Query;
import redis.clients.jedis.search.Schema;
import redis.clients.jedis.search.SearchResult;
import redis.clients.jedis.search.Document;

public class Lab3 {
    private class Product {
        private int id;
        private String gender;
        private String[] season;
        private String description;
        private double price;
        private String city;
        private String coords;

        public Product(int id, String gender, String[] season, 
            String description, double price, String city, String coords) {
                this.id = id;
                this.gender = gender;
                this.season = season;
                this.description = description;
                this.price = price;
                this.city = city;
                this.coords = coords;
        }
    }

    public void run(JedisPooled client) { 
        System.out.println("\n*** Lab 3 - Index Creation ***");
        try {client.ftDropIndex("idx1");} catch(Exception e) {};
        
        Schema schema = new Schema().addNumericField("$.id")
            .addTagField("$.gender").as("gender")
            .addTagField("$.season.*").as("season")
            .addTextField("$.description", 1.0).as("description")
            .addNumericField("$.price").as("price")
            .addTextField("$.city",1.0).as("city")
            .addGeoField("$.coords").as("coords");
        IndexDefinition rule = new IndexDefinition(IndexDefinition.Type.JSON)
            .setPrefixes(new String[]{"product:"});
        client.ftCreate("idx1", IndexOptions.defaultOptions().setDefinition(rule), schema);

        System.out.println("\n*** Lab 3 - Data Loading ***");
        Product prod15970 = new Product(15970, "Men", new String[]{"Fall", "Winter"}, "Turtle Check Men Navy Blue Shirt",
                                34.95, "Boston", "-71.057083, 42.361145");
        Product prod59263 = new Product(59263, "Women", new String[]{"Fall", "Winter", "Spring", "Summer"}, "Titan Women Silver Watch",
                                129.99, "Dallas", "-96.808891, 32.779167");
        Product prod46885 = new Product(46885, "Boys", new String[]{"Fall"}, "Ben 10 Boys Navy Blue Slippers",
                                45.99, "Denver", "-104.991531, 39.742043");
                     
        Gson gson = new Gson();
        client.jsonSet("product:15970", gson.toJson(prod15970));   
        client.jsonSet("product:59263", gson.toJson(prod59263));  
        client.jsonSet("product:46885", gson.toJson(prod46885));   
         
        System.out.println("\n*** Lab 3 - Retrieve All ***");
        Query q = new Query("*");
        SearchResult res = client.ftSearch("idx1", q);
        List<Document> docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }
            
        System.out.println("\n*** Lab 3 - Single Term Text ***");
        q = new Query("@description:Slippers");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }
  
        System.out.println("\n*** Lab 3 - Exact Phrase Text ***");
        q = new Query("@description:(\"Blue Shirt\")");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }

        System.out.println("\n*** Lab 3 - Numeric Range ***");
        q = new Query("@price:[40,130]");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }
    
        System.out.println("\n*** Lab 3 - Tag Array ***");
        q = new Query("@season:{Spring}");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }
        
        System.out.println("\n*** Lab 3 - Logical AND ***");
        q = new Query("@price:[40, 100] @description:Blue");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }   
      
        System.out.println("\n*** Lab 3 - Logical OR ***");
        q = new Query("(@gender:{Women})|(@city:Boston)");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }       

        System.out.println("\n*** Lab 3 - Negation ***");
        q = new Query("-(@description:Shirt)");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }
        
        System.out.println("\n*** Lab 3 - Prefix ***");
        q = new Query("@description:Nav*");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }
 
        System.out.println("\n*** Lab 3 - Suffix ***");
        q = new Query("@description:*Watch");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }
            
        System.out.println("\n*** Lab 3 - Fuzzy ***");
        q = new Query("@description:%wavy%");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }

        System.out.println("\n*** Lab 3 - Geo ***");
        q = new Query("@coords:[-104.800644 38.846127 100 mi]");
        res = client.ftSearch("idx1", q);
        docs = res.getDocuments();
        for (Document doc : docs) {
            System.out.println(doc);
        }
    }
}

File: Real-Time-Query-Workshop/dotnet/code/lab3.cs
/**
 * @fileoverview Basic Search operations with the redis client
 * @maker Joey Whelan
 */
using StackExchange.Redis;
using NRedisStack;
using NRedisStack.RedisStackCommands;
using NRedisStack.Search;
using NRedisStack.Search.Literals.Enums;

namespace SearchWorkshop 
{
    public class Lab3 
    { 
        public void Run(IDatabase db)
        {
            Console.WriteLine("\n*** Lab 3 - Index Creation ***");
            ISearchCommands ft = db.FT();
            try {ft.DropIndex("idx1");} catch {};
            ft.Create("idx1",   new FTCreateParams().On(IndexDataType.JSON)
                                                    .Prefix("product:"),
                                new Schema().AddNumericField(new FieldName("$.id", "id"))
                                            .AddTagField(new FieldName("$.gender", "gender"))
                                            .AddTagField(new FieldName("$.season.*", "season"))
                                            .AddTextField(new FieldName("$.description", "description"))
                                            .AddNumericField(new FieldName("$.price", "price"))
                                            .AddTextField(new FieldName("$.city", "city"))
                                            .AddGeoField(new FieldName("$.coords", "coords")));

            Console.WriteLine("\n*** Lab 3 - Data Loading ***");
            IJsonCommands json = db.JSON();
            json.Set("product:15970", "$", new {
                id = 15970, 
                gender = "Men", 
                season = new[] {"Fall", "Winter"}, 
                description = "Turtle Check Men Navy Blue Shirt", 
                price = 34.95, 
                city = "Boston", 
                coords = "-71.057083, 42.361145"
            });
            json.Set("product:59263", "$", new {
                id = 59263, 
                gender = "Women", 
                season = new[] {"Fall", "Winter", "Spring", "Summer"},
                description = "Titan Women Silver Watch", 
                price = 129.99, 
                city = "Dallas", 
                coords = "-96.808891, 32.779167"
            });
            json.Set("product:46885", "$", new {
                id = 46885, 
                gender = "Boys", 
                season = new[] {"Fall"}, 
                description = "Ben 10 Boys Navy Blue Slippers", 
                price = 45.99, 
                city = "Denver", 
                coords = "-104.991531, 39.742043"
            });

            Console.WriteLine("\n*** Lab 3 - Retrieve All ***");
            foreach (var doc in ft.Search("idx1", new Query("*")).Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }
      
            Console.WriteLine("\n*** Lab 3 - Single Term Text ***");
            foreach (var doc in ft.Search("idx1", new Query("@description:Slippers"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }

            Console.WriteLine("\n*** Lab 3 - Exact Phrase Text ***");
            foreach (var doc in ft.Search("idx1", new Query("@description:(\"Blue Shirt\")"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }

            Console.WriteLine("\n*** Lab 3 - Numeric Range ***");
            foreach (var doc in ft.Search("idx1", new Query("@price:[40,130]"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }        

            Console.WriteLine("\n*** Lab 3 - Tag Array ***");
            foreach (var doc in ft.Search("idx1", new Query("@season:{Spring}"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }          

            Console.WriteLine("\n*** Lab 3 - Logical AND ***");
            foreach (var doc in ft.Search("idx1", new Query("@price:[40, 100] @description:Blue"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }          

            Console.WriteLine("\n*** Lab 3 - Logical OR ***");
            foreach (var doc in ft.Search("idx1", new Query("(@gender:{Women})|(@city:Boston)"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }         

            Console.WriteLine("\n*** Lab 3 - Negation ***");
            foreach (var doc in ft.Search("idx1", new Query("-(@description:Shirt)"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            } 

            Console.WriteLine("\n*** Lab 3 - Prefix ***");
            foreach (var doc in ft.Search("idx1", new Query("@description:Nav*"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }

            Console.WriteLine("\n*** Lab 3 - Suffix ***");
            foreach (var doc in ft.Search("idx1", new Query("@description:*Watch"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }

            Console.WriteLine("\n*** Lab 3 - Fuzzy ***");
            foreach (var doc in ft.Search("idx1", new Query("@description:%wavy%"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }

            Console.WriteLine("\n*** Lab 3 - Geo ***");
            foreach (var doc in ft.Search("idx1", new Query("@coords:[-104.800644 38.846127 100 mi]"))
                                .Documents.Select(x => x["json"]))
            {
                Console.WriteLine(doc);
            }
        }
    }
}